File: document/Algorithm/sort/java/BubbleSort.java
/*
Bubble Sort
- Stable: Yes
- In-place: Yes
- Time: O(n^2) average/worst, O(n) best
- Space: O(1)
*/
public class BubbleSort {
    static void printArr(int[] a) {
        for (int i = 0; i < a.length; i++) {
            if (i > 0) System.out.print(" ");
            System.out.print(a[i]);
        }
        System.out.println();
    }
    public static void sort(int[] a) {
        int n = a.length;
        for (int i = 0; i < n - 1; i++) {
            boolean swapped = false;
            for (int j = 0; j < n - 1 - i; j++) {
                if (a[j] > a[j + 1]) {
                    int t = a[j]; a[j] = a[j + 1]; a[j + 1] = t;
                    swapped = true;
                }
            }
            if (!swapped) break;
        }
    }
    public static void main(String[] args) {
        int[] a = {5,1,4,2,8};
        sort(a);
        printArr(a);
    }
}

File: document/Algorithm/sort/python/bubble_sort.py
"""
Bubble Sort
- Stable: Yes
- In-place: Yes
- Time: O(n^2) average/worst, O(n) best (already sorted)
- Space: O(1)
"""

def bubble_sort(arr):
    n = len(arr)
    # Traverse the list n-1 times
    for i in range(n - 1):
        swapped = False  # Track if any swap happened in this pass
        # Last i elements are already in place
        for j in range(0, n - 1 - i):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
        # Early exit if no swaps => already sorted
        if not swapped:
            break
    return arr

if __name__ == "__main__":
    data = [5, 1, 4, 2, 8]
    print("before:", data)
    print("after :", bubble_sort(data))
